### PR TITLE
Process trait improvements

### PR DESCRIPTION
**Pull request by Ioan-Cristian**
### Pull Request Overview

This pull request brings little improvements to the `Process` trait.

### Testing Strategy

This pull request was tested by running `libtock-c` applications on STM32F429ZI Nucleo-144.

### TODO or Help Wanted

No help needed.

### Documentation Updated

- [x] Added required documentation comments.

### Formatting

- [x] Ran `make prepush`.

### CODE REVIEW

File: kernel/src/process.rs [hunk 1]
```diff
@@ -283,9 +293,8 @@ pub trait Process {
     fn short_app_id(&self) -> ShortID;
 
     /// Returns the version number of the binary in this process, as specified
-    /// in a TBF Program Header; if the Userspace Binary only has a TBF Main
-    /// Header, returns 0.
-    fn binary_version(&self) -> u32;
+    /// in a TBF Program Header; if the binary has no version assigned, return [None]
+    fn binary_version(&self) -> Option<BinaryVersion>;
```
Comment (Ioan-Cristian) [inline kernel/src/process.rs:297]: Previously, 0 marked the absence of the binary version. This approach works, but `Option` is the Rust way of doing this in my opinion.

Using a `u32` for the binary version is too powerful: several operations defined on the integer type don't make sense, like addition or division. Or do they?

In the end, NonZeroU32 has been chosen as the underlying type of the new `BinaryVersion` type. This allows code generation optimizations. In my tests, the new version compiles exactly the same to the old code.

File: kernel/src/process.rs [hunk 2]
```diff
@@ -612,7 +621,7 @@ pub trait Process {
         driver_num: usize,
         size: usize,
         align: usize,
-    ) -> bool;
+    ) -> Result<(), ()>;
```
Comment (Ioan-Cristian) [inline kernel/src/process.rs:624]: `Result` is the Rust way of signaling a function may fail. Additionally, `Result` triggers a warning when the return value is unused. From my little experiments, the new signature is compiled to the same code as the old one.

File: kernel/src/process.rs [hunk 3]
```diff
@@ -625,14 +634,14 @@ pub trait Process {
...
-    /// If successful, return a Some() with an identifier that can be used with
+    /// If successful, return a Ok() with an identifier that can be used with
     /// `enter_custom_grant()` to get access to the memory and the pointer to
     /// the memory which must be used to initialize the memory.
     fn allocate_custom_grant(
         &self,
         size: usize,
         align: usize,
-    ) -> Option<(ProcessCustomGrantIdentifier, NonNull<u8>)>;
+    ) -> Result<(ProcessCustomGrantIdentifier, NonNull<u8>), ()>;
```
Comment (Ioan-Cristian) [inline kernel/src/process.rs:644]: This follows the pattern of returning a `Result` as `allocate_grant`.

**Review by bradjc — DISMISSED** [review]
I like this change.

File: kernel/src/process.rs [hunk 4]
```diff
@@ -272,6 +273,15 @@ pub enum StoppedExecutingReason {
     KernelPreemption,
 }
 
+#[derive(PartialEq, Eq, PartialOrd, Ord)]
+pub struct BinaryVersion(NonZeroU32);
```
Comment (alevy) [inline kernel/src/process.rs:277]: Please document this type. A simple sentence describing it's purpose or meaning would be sufficient.

Comment (alevy) [inline kernel/src/process.rs:277]: I believe as it stands there is no way to read the actual version, only compare it with other versions. Perhaps just make the inner field `pub`?

```suggestion
pub struct BinaryVersion(pub NonZeroU32);
```
```

**Review by alevy — CHANGES_REQUESTED** [review]

File: kernel/src/process.rs (hunk 4 above)
Comment (Ioan-Cristian) [reply kernel/src/process.rs:278]: I think there is no need to access the inner value directly. The Tock kernel only needs to compare two different binary versions. For debug purposes, I derived the `Debug` trait.

Comment (Ioan-Cristian) [reply kernel/src/process.rs:278]: Not quite a sentence, but added a very brief description. I think it suffices.

Comment (alevy) [reply kernel/src/process.rs:278]: OK, fine then to leave as-is. If we ever need to access it, we can add it.

**Review by alevy — APPROVED** [review]

**Pull request merged**